## Ομάδα 

Στοιχεία Μέλους 1: Δόντη Ειρήνη 03119839

# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα εργαστείτε μόνο με το αρχείο movies_metadata.csv, ενώ στο μέρος τρία θα δουλέψετε και με τα αρχεία των αξιολογήσεων.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του movies_metadata.csv, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής 
14. Ίδια δεκαετία 

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.). Το συγκεκριμένο μέρος εργασίας είναι προπαρασκευαστικό οπότε όσο καλύτερα και πλουσιότερα τα ερωτήματα που θα φτιάξετε σε αυτό το μέρος τόσο καλύτερη θα είναι η απόδοσή των συστημάτων συστάσεων των επόμενων ερωτημάτων.

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata).
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv. 

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας. 


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής. 

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα, αν δύο ταινίες μοιάζουν ενισχύουμε το βάρος που προσθέτουμε από το να μοιάζουν λιγότερο) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία (διαφορετικό είναι ο χρήστης να έχει βάλει 5/5 ή 1/5 σε μια ταινία από 3/5). 

2. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη ή όχι και θα μετράμε το πόσο καλά τα πήγε το σύστήμά μας με βάση ορισμένες μετρικές.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο test_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θεωρούμε οτι προτείνεται στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3 στο test_ratings. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να προβλέψετε εάν θα αρέσει στο χρήστη ή όχι ώστε να την προτείνετε. 

ΠΡΟΣΟΧΗ! Η "βαθμολογία" που θα υπολογίσετε για κάθε ταινία κατά το training δεν είναι απαραίτητα πρόβλεψη της βαθμολογίας που θα έβαζε ο χρήστης. 

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων από ελάχιστες, λίγες μέχρι και πολλές (π.χ. 3, 5, 10, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set). 

Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100. Έτσι για παράδειγμα αν ο αλγόριθμος τα πηγαίνει πολύ καλά για τρεις ταινίες και στη συνέχεια το σκορ βελτιώνεται ελάχιστα τότε αυτός ενδείκνυται για ένα σύστημα συστάσεων για νέους χρήστες όπου δεδομένου λίγων ταινιών ο αλγόριθμος είναι σε θέση να το να προτείνει καλές συστάσεις. Ενώ αν η καλύτερη απόδοση του αλγορίθμου σας είναι βέλτιστη με περισσότερες ταινίες π.χ. 50 τότε αυτός ο αλγόριθμος συστάσεων ενδείκνυται για παλιούς χρήστες με πολλές αξιολόγησεις. 

---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί. 

```
from google.colab import drive
drive.mount('/content/drive')
import os
os.listdir('/gdrive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/gdrive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά. 

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


---
#Κώδικας για κατασκευή περιβάλλοντος εργασίας

##Μόνο για Google Colab

Κώδικας για να γίνει Mount to Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import os
path="" # με ευκόλο τρόπο μπορείτε να αλλάξετε το path που είναι αποθηκευμένα τα αρχεία σας

Mounted at /content/drive


In [ ]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [ ]:
import pandas as pd
from pyswip import Prolog

In [ ]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd 
path = '/content/drive/MyDrive/'
# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv(path + "movies_metadata.csv") 
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data 
data.head()

Unnamed: 0     budget                                    genres  \
0           0  237000000  Action|Adventure|Fantasy|Science Fiction   
1           1  300000000                  Adventure|Fantasy|Action   
2           2  245000000                    Action|Adventure|Crime   
3           3  250000000               Action|Crime|Drama|Thriller   
4           4  260000000          Action|Adventure|Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                       plot_keywords  language  \
0  culture clash|future|space war|space colony|so...   English   
1  ocean|drug abuse|exotic island|east india trad...   English   
2  spy|based on novel|secret agent|sequel|mi6|bri...  Français   
3  dc comics|crime fighter|terrorist|secret ident...   English   
4  based on novel|mars|medallion|space travel|pri...   English   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ...  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   
3  Following the death of District Attorney Harve...  112.312950  ...   
4  John Carter is a war-weary, former military ca...   43.926995  ...   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                movie_title vote_average  num_voted_users  \
0                                    Avatar          7.2            11800   
1  Pirates of the Caribbean: At World's End          6.9             4500   
2                                   Spectre          6.3             4466   
3                     The Dark Knight Rises          7.6             9106   
4                               John Carter          6.1             2124   

  title_year                   country      director_name     actor_1_name  \
0     2009.0  United States of America      James Cameron      Zoe Saldana   
1     2007.0  United States of America     Gore Verbinski    Orlando Bloom   
2     2015.0            United Kingdom         Sam Mendes  Christoph Waltz   
3     2012.0  United States of America  Christopher Nolan    Michael Caine   
4     2012.0  United States of America     Andrew Stanton     Lynn Collins   

       actor_2_name       actor_3_name  
0  Sigourney Weaver       Stephen Lang  
1   Keira Knightley  Stellan Skarsgård  
2       Léa Seydoux      Ralph Fiennes  
3       Gary Oldman      Anne Hathaway  
4   Samantha Morton       Willem Dafoe  

[5 rows x 27 columns]

In [ ]:
def clean_text(text):
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  return text

In [ ]:
# for c in data.columns:
#   print(c)

Unnamed: 0
budget
genres
homepage
id
plot_keywords
language
original_title
overview
popularity
production_companies
production_countries
release_date
gross
duration
spoken_languages
status
tagline
movie_title
vote_average
num_voted_users
title_year
country
director_name
actor_1_name
actor_2_name
actor_3_name


In [ ]:
#create World
#Ορίζουμε τον κόσμο μας
prolog = Prolog()

#Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκέυσουμε
#αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}
color = 'RGB'
for row in data.itertuples(index=True, name='Pandas'):
  movie_title = clean_text(getattr(row, 'movie_title'))  
  # genres = clean_text(getattr(row, 'genres'))  
  plot_keywords = clean_text(getattr(row, 'plot_keywords'))
  popularity = clean_text(str(getattr(row, 'popularity')))
  duration= clean_text(str(getattr(row, 'duration')))  
  vote_average= clean_text(str(getattr(row, 'vote_average')))  
  num_voted_users= clean_text(str(getattr(row, 'num_voted_users')))  
  title_year= clean_text(str(getattr(row, 'title_year')))  
  director_name= clean_text(getattr(row, 'director_name'))  
  production_companies = clean_text(getattr(row, 'production_companies'))
  production_countries = clean_text(getattr(row, 'production_countries'))
  language = clean_text(getattr(row, 'language'))


  if movie_title != 'UNK':
    if director_name != 'UNK':
      literals.append("director('"+ movie_title +"','" + director_name + "')")
    if popularity != 'UNK':
      literals.append("popularity('"+ movie_title +"','" + popularity + "')")
    if num_voted_users != 'UNK':
      literals.append("num_voted_users('"+ movie_title +"','" + num_voted_users + "')")
    if title_year != 'UNK':
      literals.append("title_year('"+ movie_title +"','" + title_year + "')")
    if vote_average != 'UNK':
      literals.append("vote('"+ movie_title +vote_average+"')")
    if duration != 'UNK':
       literals.append("duration('"+ movie_title +duration+"')")
    if production_companies != 'UNK':
      literals.append("production_companies('"+ movie_title +"','" + production_companies + "')")
    if production_countries != 'UNK':
      literals.append("production_countries('"+ movie_title +"','" + production_countries + "')")
    if language != 'UNK':
      literals.append("language('"+ movie_title +"','" + language + "')")


    for plot_keyword in plot_keywords.split("|"):
      literals.append("keyword('"+ movie_title +"','"+ plot_keyword +"')")
      if plot_keyword == "black and white":
        color = "Black_and_White"
    literals.append("color('"+ movie_title +"','" + color + "')")

    for genre in getattr(row, 'genres').split("|"):
      literals.append("genre('"+ movie_title +"','"+ genre +"')")

    for actor in [getattr(row, 'actor_1_name'), getattr(row, 'actor_2_name'), getattr(row, 'actor_2_name')]:
      literals.append("actors('"+ movie_title +"','"+ actor.replace("'",'"') +"')")

#Η Prolog θέλει τα κατηγορήματά της με την σειρά 
literals.sort()
for literal in literals:
  prolog.assertz(literal)
#   print (literal +'.')

#Επίσης μπορούμε να κάνουμε consult ένα έτοιμο αρχείο στον κόσμο όπως παρακάτω
prolog.consult(path + "db.pl")

In [ ]:

#1) Common Subject                                 
prolog.assertz('(common_sub(X, Y) :-  genre(X,A), genre(Y,A), genre(X,B), genre(Y,B), genre(X,C), genre(Y,C), genre(X,D), genre(Y,D), X\=Y, is_set([A,B,C,D]))') ## A\=B, A\=C, A\=D, B\=C, B\=D, C\=D,
# prolog.assertz('(common_sub(X, Y) :-  genre(X,D), genre(Y,D), X\=Y)') ## A\=B, A\=C, A\=D, B\=C, B\=D, C\=D,

#2) Less common Subject                                
prolog.assertz('(less_common_sub(X, Y) :-  genre(X,A), genre(X,B), genre(X,C), genre(Y,A), genre(Y,B), genre(Y,C), X\=Y, A\= B, A\= C, B\=C, not(common_sub(X, Y)) )')


#3) Even Less Common Subject                           
prolog.assertz('(even_less_common_sub(X, Y) :-  genre(X,A), genre(Y,A), X\=Y, not(less_common_sub(X, Y)))')

#4) Common Director                                    
prolog.assertz('(common_director(X, Y) :-  director(X,A), director(Y,A), X\=Y )')

#5) Same Plot                                          
prolog.assertz('(same_plot4(X, Y) :-  keyword(X,A), keyword(Y,A), keyword(X,B), keyword(Y,B), keyword(X,C), keyword(Y,C), keyword(X,D), keyword(Y,D), A\=B, A\=C, A\=D, B\=C, B\=D, C\=D, X\=Y)')
prolog.assertz('(same_plot3(X, Y) :-  keyword(X,A), keyword(X,B), keyword(X,C), keyword(Y,A), keyword(Y,B), keyword(Y,C), X\=Y, A\= B, A\= C, B\=C, not(same_plot4(X, Y)) )')

#6) Common Plot                                 
prolog.assertz('(same_plot2(X, Y) :-  keyword(X,A), keyword(X,B), keyword(Y,A), keyword(Y,B), X\=Y , A\= B, not(same_plot3(X, Y)))')
prolog.assertz('(same_plot1(X, Y) :-  keyword(X,A), keyword(X,B), X\=Y , A \= B, not(same_plot2(X, Y)))')

#7) Three common actors                               
prolog.assertz('(common_actors3(X, Y) :-  actors(X,A), actors(X,B), actors(X,C), actors(Y,A), actors(Y,B), actors(Y,C), X\=Y, A \= B, A \= C, B \= C)')

#8) Two common actors      
prolog.assertz('(common_actors2(X, Y) :-  actors(X,A), actors(X,B), actors(Y,A), actors(Y,B), X\=Y, A \= B, not(common_actors3(X, Y)) )')

#9) One common actor                                 
prolog.assertz('(common_actors1(X, Y) :-  actors(X,A), actors(Y,A), X\=Y, not(common_actors2(X, Y)) )')

#10) Common Language                                 
prolog.assertz('(common_language(X, Y) :- language(X,D), language(Y,D), X \= Y)')

#11) Same color             
prolog.assertz('(same_color(X,Y) :- color(X, C), color(Y, C), X \= Y )')

#12) Common studio
prolog.assertz('(common_studio(X,Y) :- production_companies(X,A), production_companies(Y,A), X \= Y)')

#13) Common pruduction country
prolog.assertz('(common_prod_country(X,Y) :- production_countries(X,A), production_countries(Y,A), X \= Y)')

#14) Same decade
prolog.assertz('(same_decade(X,Y) :- title_year(X,A), title_year(Y,B), ((A mod 100-A mod 10)=(B mod 100-B mod 10)), X \= Y)')



In [ ]:
similarities = ["common_sub('","less_common_sub('","even_less_common_sub('", "common_director('","same_plot4('",
      "same_plot3('","same_plot2('", "same_plot1('", "common_actors3('",
      "common_actors2('","common_actors1('","common_language('", "same_color('", "common_studio('","common_prod_country('", "same_decade('"]

for x in similarities: 
  querry = prolog.query(x + 'Spectre' +"',M)")
  result = set()
  for solution in querry:
    result.add(solution["M"])
  print(x.split("(")[0],list(set(result)))

common_sub ['Krrish', 'Two Brothers', 'Smillas Sense of Snow', 'Madison', 'Escape from Planet Earth', 'Were No Angels', 'Captive', 'My Super Ex-Girlfriend', 'Brigham City', 'House of Flying Daggers', 'The Cottage', 'Donkey Punch', 'The Shadow', 'Munich', 'Blackhat', 'After the Sunset', 'Batman', 'Tomorrow Never Dies', 'Rat Race', 'Unnatural', 'Jackie Brown', 'Breakdown', 'The Boondock Saints', 'The Stepford Wives', 'Legend', 'Highlander', 'Mr. Peabody & Sherman', 'The Road', 'Screwed', 'Virgin Territory', 'Snow White and the Huntsman', 'The Interview', 'Immortals', 'Love Me Tender', 'Quest for Fire', 'Return to the Blue Lagoon', 'Malibus Most Wanted', '127 Hours', 'Oceans Twelve', 'Lucky Numbers', 'Ink', 'Observe and Report', 'Subway', 'Back to the Future Part III', 'Cars 2', 'Safe House', 'Star Trek II: The Wrath of Khan', 'Under Siege 2: Dark Territory', '55 Days at Peking', 'The Good Dinosaur', 'Legend of a Rabbit', 'Daybreakers', 'Project X', 'Tsotsi', 'The Big Hit', 'The Mask of Z

In [ ]:
similarities = ["common_sub('","less_common_sub('","even_less_common_sub('", "common_director('","same_plot4('",
      "same_plot3('","same_plot2('", "same_plot1('", "common_actors3('",
      "common_actors2('","common_actors1('","common_language('", "same_color('", "common_studio('","common_prod_country('", "same_decade('"]

for x in similarities: 
  querry = prolog.query(x + 'Man of Steel' +"',M)")
  result = set()
  for solution in querry:
    result.add(solution["M"])
  print(x.split("(")[0],list(set(result)))

common_sub ['Krrish', 'Two Brothers', 'Smillas Sense of Snow', 'Madison', 'Escape from Planet Earth', 'Quarantine', 'My Super Ex-Girlfriend', 'House of Flying Daggers', 'The Shadow', 'Quest for Camelot', 'Munich', 'After the Sunset', 'Batman', 'Tomorrow Never Dies', 'Charly', 'Rat Race', 'Unnatural', 'Breakdown', 'The Boondock Saints', 'The Stepford Wives', 'Multiplicity', 'Highlander', 'Mr. Peabody & Sherman', 'The Road', 'Screwed', 'Virgin Territory', 'Snow White and the Huntsman', 'The Interview', 'Immortals', 'Love Me Tender', 'Quest for Fire', 'Return to the Blue Lagoon', 'Gravity', 'The Mist', '127 Hours', 'Lucky Numbers', 'A Scanner Darkly', 'Ink', 'The Muppet Christmas Carol', 'Subway', 'Back to the Future Part III', 'Baby Geniuses', 'Cars 2', 'Safe House', 'Cirque du Soleil: Worlds Away', 'Star Trek II: The Wrath of Khan', 'Under Siege 2: Dark Territory', '55 Days at Peking', 'Like Mike', 'The Good Dinosaur', 'Legend of a Rabbit', 'Daybreakers', 'Skyline', 'The Invention of Ly

**Σχολιασμός για μέρος 1**

Αφού μελετήσαμε προσεκτικά το αρχείο "movies_metadata.csv" και εντοπίσαμε τις πιο σημαντικές (κατά την γνώμη μας) στήλες, κατασκευάσαμε τον κόσμο της prolog (ερώτημα 1α). 
Στην συνέχεια, υλοποιήσαμε τα ζητούμενα queries (ερώτημα 1β), τα οποία χρησιμοποιούνται για να βρίσκει στο σύστημα ποιες ταινίες παρουσιάζουν μεταξύ τους ομοιότητες, και τι είδους είναι αυτές οι ομοιότητες.

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4. 


In [ ]:
########## 5 ##########
prolog.assertz('(find_sim_5(X,Y):- less_common_sub(X,Y), same_plot2(X,Y), common_actors1(X,Y), common_language(X, Y), common_studio(X,Y))')

########## 4 ##########
prolog.assertz('(find_sim_4(X,Y):- common_sub(X, Y), same_plot3(X, Y),  common_studio(X,Y), same_decade(X,Y))')

########## 3 ##########
prolog.assertz('(find_sim_3(X,Y):- less_common_sub(X, Y), same_plot2(X, Y))')

########## 2 ##########
prolog.assertz('(find_sim_2(X,Y):-  common_sub(X, Y), same_plot1(X, Y), common_director(X, Y);same_decade(X,Y))')

########## 1 ##########
prolog.assertz('(find_sim_1(X,Y):- common_sub(X, Y))')



In [ ]:
def check_query(qu):
    s = set()
    for res in qu:
      m = res['M'] 
      if m not in s:
        s.add(res['M'])
    return s


def simple_recommender(movie):
    recommended_movies = []
    visited = set()
    how_similar = 0
    # print("5...")
    q = prolog.query("find_sim_5('" + movie +"',M)")
    s = check_query(q)
    s5 = set()
    if(s):
      how_similar = 5
      s5 = s
      if s5 not in visited:
        recommended_movies.append((how_similar, s5))
    
    # print("4...")
      
    q = prolog.query("find_sim_4('" + movie +"',M)")
    s = check_query(q)
    s4 = set()
    if(s):
      how_similar = 4
      s4 = s
      if s4 not in visited:
        recommended_movies.append((how_similar, s4))
    
    # print("3...")
    
    q = prolog.query("find_sim_3('" + movie +"',M)")
    s = check_query(q)
    s3 = set()
    if(s):
      how_similar = 3
      s3 = s
      if s3 not in visited:
        recommended_movies.append((how_similar, s3))

    # print("2...")
      
    q = prolog.query("find_sim_2('" + movie +"',M)")
    s = check_query(q)
    s2 = set()
    if(s):
      how_similar = 2
      s2 = s
      if s2 not in visited:
        recommended_movies.append((how_similar, s2))
 
    
    # print("1...")
     
    q = prolog.query("find_sim_1('" + movie +"',M)")
    s = check_query(q)
    s1 = set()
    if(s):
      how_similar = 1
      s1 = s
      if s1 not in visited:
        recommended_movies.append((how_similar, s1))
    

    recommended_movies_list = []      
    for idx in recommended_movies:
      for movie in idx[1]:
        recommended_movies_list.append(movie)

    return recommended_movies_list
  

In [ ]:
movie = "Pirates of the Caribbean: The Curse of the Black Pearl"
simple_recommender(movie)[:5]

['Elf',
 'Rugrats Go Wild',
 'The Life of David Gale',
 'The Haunted Mansion',
 'The Last Samurai']

In [ ]:
movie2 = "John Carter"
simple_recommender(movie2)[:5]

['Black November',
 'The Sweeney',
 'Holy Motors',
 'Fly Me to the Moon',
 'Arbitrage']

**Σχολιασμός για μέρος 2**

Δημιουργούμε ένα recommendation system στο οποίο κατασκευάζουμε queries που θα επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες ώστε αυτές στην συνέχεια να προταθούν στον χρήστη. Ως είσοδος, λαμβάνεται μία ταινία και επιστρέφονται οι πιο όμοιες με αυτή, με φθίνουσα σειρά ομοιότητας (αρχικά οι ταινίες με δείκτη ομοιότητας 5, 4,... κ.λ.π.). Η σειρά εμφάνισης των ταινιών με ίδιο δείκτη ομοιότητας, εξαρτάται από τη σειρά εμφάνισης που καθορίζεται από την prolog. Σε περίπτωση που η ταξινόμηση σε κλάσεις ομοιότητας εκφράζει μεγαλύτερη ομοιότητα,
ορίζουμε πιο αυστηρούς περιορισμούς και κριτήρια όπως ίδια πλοκή, θέμα, ηθοποιοί.
Στην αντίθετη περίπτωση, που η ταξινόμηση σε κλάσεις ομοιότητας εκφράζει μικρότερη ομοιότητα, ορίζουμε λιγότερο αυστηρούς περιορισμούς και κριτήρια όπως λιγότερο ίδια πλοκή, παρόμοιο θέμα ή χρώμα κλπ.
 Το recommender system που παρουσιάστηκε παραπάνω λειτουργεί ορθά, καθώς οι sequel ταινίες αποδίδονται ως περισσότερο όμοιες σε σχέση με τις υπόλοιπες ταινίες που αποτιμώνται συναφείς με τη δοθείσα ως είσοδο.

# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [ ]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {i:i + 1 for i in range(1)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = simple_recommender(movie)

        for similar_movie in similar_movies:
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[0]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[0] # το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
    return movie_score


# Αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict.
# Έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0.
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

Οι παραπάνω εκπαιδεύουν, τεστάρουν και μετρούν την απόδοσή του συστήματος στάσεων μας. Για την εκπαίδευση του συστήματος μπορούμε, κάθε φορά, να χρησιμοποιήσουμε ένα τυχαίο υποσύνολο του training set. Όμως είναι πιθανό το υποσύνολο των ταινιών αυτών να επηρεάζει τα αποτελέσματα στο training set. Για παράδειγμα από 3 ταινίες στις 10 μπορούν για ένα πείραμα τα αποτελέσματα μας αν είναι ίδια και αυτό να μην οφείλεται στο γεγονός ότι ο ταξινομητής μας τα πηγαίνει καλά στις 3 ταινίες αλλά στο γεγονός ότι οι υπόλοιπες 7 είναι τέτοιες ταινίες που δεν μας βοηθούν καθόλου και αν είχαμε επιλέξει διαφορετικές 3 ταινίες να τα πηγαίναμε χάλια. Οπότε προτείνουμε να τρέξετε κάθε φορά έναν αριθμό πειραμάτων για κάθε υποσύνολο ταινιών π.χ. 10 πειράματα με 3 ταινίες, 10 πειράματα για 20 ταινίες κ.ο.κ. και να κρατήσετε σαν τελικό σκορ το μέσο όρο όλων των πειραμάτων. Έτσι τα αποτελέσματα των αγγλικών σας θα είναι πιο αντικειμενικά.

In [ ]:
train_ratings = pd.read_csv(path + "train_ratings.csv")
test_ratings = pd.read_csv(path + "test_ratings.csv")

In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.51673565836327
recall: 0.7402777777777778
f1: 0.5780523844932199


In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5132054662220917
recall: 0.8902777777777778
f1: 0.6445002684238375


In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 45)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5112478853064859
recall: 0.9819444444444445
f1: 0.6723786687054891


**Σχολιασμός για μέρος 3**

Στο συγκεκριμένο μέρος χρησιμοποιούμε τις μετρικές precision, recall, f1 για την επίβλεψη-μέτρηση της απόδοσης του συστήματός μας.
Το precision μας πληροφορεί για την ακρίβεια του συστήματος και συγκεκριμένα μας ενημερώνει σχετικά με το πλήθος των ταινιών που λανθασμένα προβλέφθηκαν ως προτεινόμενες στον χρήστη. Παραπάνω πρατηρούμε ότι το precision κυμαίνεται περίπου γύρω από το 0.51 ή στο 51%, δηλαδή οι μισές περίπου ταινίες εσφαλμένα προτάθηκαν στον χρήστη, ενώ οι περίπου άλλες μισές προτάθηκαν ορθώς.
Το recall υπολογίζει πόσες από τις προτεινόμενες ταινίες προβλέφθηκαν σωστά. Παραπάνω παρατηρούμε ότι όσο αυξάνεται το πλήθος των ταινιών τόσο αυξάνεται η τιμή του recall (0.74 -> 0.89 -> 0.98), πράγμα το οποίο είναι λογικό αφού έχουμε περισσότερες επιλογές για να προτείνουμε.
Το f1 διατηρεί μια ισορροπία ανάμεσα στις άλλες δύο μετρικές, καθώς είναι ο μέσος τους. Παρατηρούμε ότι και στις 3 περιπτώσεις επαληθεύεται η εξίσωση 
f1=2*(precision * recall)/(precision + recall)




